src/init.cpp

```diff
@@ -1276,6 +1276,48 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
                                      chainman, *node.mempool, ignores_incoming_txs);
     RegisterValidationInterface(node.peerman.get());
 
+    // Check port numbers
+    for (const std::string& port_option : {
+        "-port",
+        "-rpcport",
+    }) {
+        if (args.IsArgSet(port_option)) {
+            int64_t port_in = args.GetIntArg(port_option, 0);
+            if (port_in <= 0 || port_in > std::numeric_limits<uint16_t>::max())
+                return InitError(InvalidPortErrMsg(port_option, ToString(port_in)));
+        }
+    }
+
+    for (const std::string& port_option : {
+        "-i2psam",
+        "-onion",
+        "-proxy",
+        "-rpcbind",
+        "-torcontrol",
+        "-whitebind",
+        "-zmqpubhashblock",
+        "-zmqpubhashtx",
+        "-zmqpubrawblock",
+        "-zmqpubsequence",
+    }) {
+        for (std::string& socket_addr : args.GetArgs(port_option)) {
+            std::string host_out;
+            uint16_t port_out{0};
+            if (!SplitHostPort(socket_addr, port_out, host_out)) {
+                return InitError(InvalidPortErrMsg(port_option, socket_addr));
+            }
+        }
+    }
+
+    for (std::string& socket_addr : args.GetArgs("-bind")) {
+        std::string host_out;
+        uint16_t port_out{0};
+        std::string bind_socket_addr = socket_addr.substr(0, socket_addr.rfind('='));
+        if (!SplitHostPort(bind_socket_addr, port_out, host_out)) {
+            return InitError(InvalidPortErrMsg("-bind", bind_socket_addr));
```

Comment:
Might be better to include the full value in the error? idk

```suggestion
            return InitError(InvalidPortErrMsg("-bind", socket_addr));
```